src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/BuildUpToDateCheck.cs

```diff
@@ -0,0 +1,308 @@
+            .Add(ResolvedAssemblyReference.SchemaName)
+            .Add(ResolvedCOMReference.SchemaName)
+            .Add(ResolvedProjectReference.SchemaName);
+
+        private static ImmutableHashSet<string> ItemSchemas => ImmutableHashSet<string>.Empty
+            .Add(AdditionalFiles.SchemaName)
+            .Add(Compile.SchemaName)
+            .Add(Content.SchemaName)
+            .Add(EmbeddedResource.SchemaName)
+            .Add(None.SchemaName);
+
+        private static ImmutableHashSet<string> ProjectPropertiesSchemas => ImmutableHashSet<string>.Empty
+            .Add(ConfigurationGeneral.SchemaName)
+            .Union(ReferenceSchemas);
+
+        private readonly IProjectSystemOptions _projectSystemOptions;
+        private readonly ConfiguredProject _configuredProject;
+        private readonly Lazy<IFileTimestampCache> _fileTimestampCache;
+
+        private IDisposable _link;
+        private IComparable _lastVersionSeen;
+
+        private bool _isDisabled = true;
+        private string _msBuildProjectFullPath;
+        private HashSet<string> _imports = new HashSet<string>();
+        private Dictionary<string, HashSet<string>> _items = new Dictionary<string, HashSet<string>>();
+        private Dictionary<string, HashSet<string>> _references = new Dictionary<string, HashSet<string>>();
+        private Dictionary<string, HashSet<string>> _outputGroups = new Dictionary<string, HashSet<string>>();
+
+        [ImportingConstructor]
+        public BuildUpToDateCheck(
+            IProjectSystemOptions projectSystemOptions,
+            ConfiguredProject configuredProject,
+            Lazy<IFileTimestampCache> fileTimestampCache)
+        {
+            _projectSystemOptions = projectSystemOptions;
+            _configuredProject = configuredProject;
+            _fileTimestampCache = fileTimestampCache;
+
+            _link = ProjectDataSources.SyncLinkTo(
+                _configuredProject.Services.ProjectSubscription.JointRuleSource.SourceBlock.SyncLinkOptions(new StandardRuleDataflowLinkOptions { RuleNames = ProjectPropertiesSchemas }),
+                _configuredProject.Services.ProjectSubscription.ImportTreeSource.SourceBlock.SyncLinkOptions(),
+                _configuredProject.Services.ProjectSubscription.SourceItemsRuleSource.SourceBlock.SyncLinkOptions(new StandardRuleDataflowLinkOptions { RuleNames = ItemSchemas }),
+                _configuredProject.Services.OutputGroups.SourceBlock.SyncLinkOptions(),
+                target: new ActionBlock<IProjectVersionedValue<Tuple<IProjectSubscriptionUpdate, IProjectImportTreeSnapshot, IProjectSubscriptionUpdate, IImmutableDictionary<string, IOutputGroup>>>>(e => OnChanged(e)),
+                linkOptions: new DataflowLinkOptions { PropagateCompletion = true });
+        }
+
+        private void OnProjectChanged(IProjectSubscriptionUpdate e)
+        {
+            if (e.ProjectChanges.TryGetValue(ConfigurationGeneral.SchemaName, out var projectChange))
```

Comment:
We have a GetPropertyOrDefault extension method that'll make this easier to read.